src/Workleap.Extensions.Mongo.Ephemeral/ReusableMongoRunnerProvider.cs

```diff
@@ -1,38 +1,90 @@
+
+        lock (this._rentedRunnersByClientName)
+        {
+            if (this._rentedRunnersByClientName.TryGetValue(clientName, out var rentedRunners))
+            {
+                rentedRunners.Add(runner);
+            }
+            else
+            {
+                this._rentedRunnersByClientName[clientName] = [runner];
+            }
+        }
+
+        return runner;
     }
 
-    public ReusableMongoRunner GetRunner(string clientName)
+    private static Lazy<MongoRunnerPool> CreateLazyMongoRunnerPool(string clientName)
     {
-        var runner = LazyRunners.GetOrAdd(clientName, CreateLazyReusableMongoRunner).Value;
-        runner.Rent(this._renterId);
-        return runner;
+        return new Lazy<MongoRunnerPool>(CreatePool);
     }
 
-    private static Lazy<ReusableMongoRunner> CreateLazyReusableMongoRunner(string clientName)
+    private static MongoRunnerPool CreatePool()
     {
-        return new Lazy<ReusableMongoRunner>(static () => new ReusableMongoRunner());
+        var options = new MongoRunnerOptions
+        {
+            Version = MongoVersion.V8,
+            UseSingleNodeReplicaSet = true,
+        };
+
+        var version = Environment.GetEnvironmentVariable("WORKLEAP_EXTENSIONS_MONGO_EPHEMERAL_VERSION")?.Trim();
+        if (int.TryParse(version, NumberStyles.Integer, CultureInfo.InvariantCulture, out var parsedVersion))
+        {
+            options.Version = parsedVersion switch
+            {
+                6 => MongoVersion.V6,
+                7 => MongoVersion.V7,
+                8 => MongoVersion.V8,
+                _ => options.Version
+            };
+        }
+
+        var edition = Environment.GetEnvironmentVariable("WORKLEAP_EXTENSIONS_MONGO_EPHEMERAL_EDITION")?.Trim();
+        if ("enterprise".Equals(edition, StringComparison.OrdinalIgnoreCase))
+        {
+            options.Edition = MongoEdition.Enterprise;
+            options.AdditionalArguments = ["--storageEngine", "inMemory"];
+        }
+
+        var connectionTimeout = Environment.GetEnvironmentVariable("WORKLEAP_EXTENSIONS_MONGO_EPHEMERAL_CONNECTIONTIMEOUT")?.Trim();
+        if (TimeSpan.TryParse(connectionTimeout, CultureInfo.InvariantCulture, out var parsedConnectionTimeout))
+        {
+            options.ConnectionTimeout = parsedConnectionTimeout;
+        }
+
+        return new MongoRunnerPool(options, maxRentalsPerRunner: 100);
     }
 
     public void Dispose()
     {
-        foreach (var lazyRunner in LazyRunners.Values)
+        lock (this._rentedRunnersByClientName)
         {
-            if (lazyRunner.IsValueCreated)
+            foreach (var (clientName, rentedRunners) in this._rentedRunnersByClientName)
             {
-                lazyRunner.Value.Return(this._renterId);
+                foreach (var rentedRunner in rentedRunners)
+                {
+                    LazyRunnerPools[clientName].Value.Return(rentedRunner);
```

Comment:
I would check the `IsValueCreated`. In case there is an exception when creating the value, it would avoid throwing here again.

Reply:
Re-reading the code, it should not happen. You can forget this comment.